File: my-folder/0022-generate-parentheses/solution.java
class Solution {
    
    List<String> res = new ArrayList<>();
    
    public List<String> generateParenthesis(int n) {
        // constrcut a string object out of a char array of length 2*n from current index 
        // and contains well-formed parentheses. 
        backtrack(new char[n*2], 0, 0, 0, n);
        return res;
    }
    
    private void backtrack(char[] arr, int currIndex, int open, int close, int max) {
        if (close == max) {
            res.add(new String(arr));
            return;
        }
        
        if (open < max) {
            arr[currIndex] = '(';
            backtrack(arr, currIndex+1, open+1, close, max);
        }
        
        if (close < open) {
            arr[currIndex] = ')';
            backtrack(arr, currIndex+1, open, close+1, max);
        }
    }
}

File: my-folder/0022-generate-parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def traverse(now_str, l, r):
            if len(now_str)==2*n:
                res.append(now_str)
            if l<n:
                traverse(now_str+"(", l+1, r)
            if r<l:
                traverse(now_str+")", l, r+1)
        traverse("", 0, 0)
        return res
        
